### PR TITLE
Audit delay script and accompanying unit tests

### LINKED ISSUE #34: implement auditDelay script
# Bounty

when capacity is [withdrawn from the bridge](https://github.com/leapdao/nervos/issues/33), it is put into an `auditDelay()` script. here the withdrawal is either canceled by the trustee within a given time, or the capacity becomes free to be claimed by the recipient.

parentBridgeTxns (4)

## Scope

- take `receiverAddress`, `trusteeAddress` and `typeHash` of bridge as args. have `unlock-time` hardcoded.
- take header dependency of creation of the auditDelay cell and header dep of one of the current blocks, check that at least `unlock-time` has passed between the two blocks, allow to unlock capacity to `receiverAddress`
- if input 0 is of `typeHash` and witness carries a sig by `trusteeAddress`, lock capacity back into bridge.
-

## Deliverables

- code
- tests

## Funding Circle
Nervos

## Bounty Owner/Gardener

<@your github username> as **<Role>**

## Roles
bounty gardener: @johannbarbie / 15%
bounty worker: name / share
bounty reviewer: name / share

## Gardener checklist

- [ ] I have attached one of the `size-` labels to this bounty
- [ ] I have submitted this bounty via http://bounty.leapdao.org/viewform
- [ ] I have added this bounty to bounty board here https://github.com/orgs/leapdao/projects/6

### PR DESCRIPTION
**Pull request by Strumster**
Resolves #34 

Built with ckb-testtool v2.2, as it added header calls.

Unit tests end with expect_err call, since assert_error_eq! call seems to be broken with 2.2.

### CODE REVIEW

File: packages/parent-bridge/tests/Cargo.toml [hunk 1]
```diff
@@ -6,6 +6,7 @@ edition = "2018"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-ckb-tool = { git = "https://github.com/jjyr/ckb-tool.git", tag = "v0.1.1" }
-ckb-testtool = { git = "https://github.com/jjyr/ckb-tool.git", tag = "v0.1.1" }
+ckb-tool = { git = "https://github.com/jjyr/ckb-tool.git", branch="master"}
```
Comment (mwaa) [inline packages/parent-bridge/tests/Cargo.toml:9]: @Strumster  Can we change this to use the official releases. Please note there may be some folder changes that would need to be changed for all contracts. Then test that nothing was broken.

https://github.com/nervosnetwork/capsule/releases

File: packages/parent-bridge/contracts/audit-delay/src/entry.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    let trustee_lock_hash = args.slice(0..32);
+    let owner_lock_hash = args.slice(32..64);
+
+    if Bytes::from(&input0_lock_hash[..]) == trustee_lock_hash {
+        return Ok(());
+    }
+
+    let input_header = load_header(0, Source::GroupInput)?;
+    let input_header_timestamp = input_header.raw().timestamp().unpack();
+    debug!("{:?}", input_header_timestamp);
```
Comment (mwaa) [inline packages/parent-bridge/contracts/audit-delay/src/entry.rs:38]: Can we remove all debug statements or have them as comments?

File: packages/parent-bridge/tests/Cargo.toml (hunk 1 above)
Comment (Strumster) [reply packages/parent-bridge/tests/Cargo.toml:9]: load_header function calls don't work in official releases of dev tool, so audit delay wouldn't function with official version. Hopefully dev ads tag to the updated version.

File: packages/parent-bridge/contracts/audit-delay/src/entry.rs (hunk 2 above)
Comment (Strumster) [reply packages/parent-bridge/contracts/audit-delay/src/entry.rs:38]: Yeah, I forgot to remove them myself.

File: packages/parent-bridge/tests/Cargo.toml (hunk 1 above)
Comment (mwaa) [reply packages/parent-bridge/tests/Cargo.toml:9]: Sure can test as is for now.

Comment (mwaa) [reply packages/parent-bridge/tests/Cargo.toml:9]: My bad i was confusing the libraries

**Comment by mwaa** [comment]
> Resolves #34
> 
> Built with ckb-testtool v2.2, as it added header calls.
> 
> Unit tests end with expect_err call, since assert_error_eq! call seems to be broken with 2.2.

@Strumster Seems there was change in the api structure for errors. Reference example test on tool repo
https://github.com/jjyr/ckb-tool/blob/master/test-contract/tests/src/tests.rs

Please make changes to all tests to have a similar structure ending with `.input_lock_script(0)`;
Example
`assert_error_eq!(err, ScriptError::ValidationFailure(7).input_lock_script(0));`

**Comment by Strumster** [comment]
> > Resolves #34
> > Built with ckb-testtool v2.2, as it added header calls.
> > Unit tests end with expect_err call, since assert_error_eq! call seems to be broken with 2.2.
> 
> @Strumster Seems there was change in the api structure for errors. Reference example test on tool repo
> https://github.com/jjyr/ckb-tool/blob/master/test-contract/tests/src/tests.rs
> 
> Please make changes to all tests to have a similar structure ending with `.input_lock_script(0)`;
> Example
> `assert_error_eq!(err, ScriptError::ValidationFailure(7).input_lock_script(0));`

Done. Note that this doesn't function with devtool 2.2, hopefully it will be fixed in a new release.

**Comment by mwaa** [comment]
@TheReturnOfJan Cancel payout is missing in this functionality. Is there a way to add it without implementing trustee?

**Review by mwaa — APPROVED** [review]
Cancel Payout forked out to new issue #72

**Pull request merged**